src/map.cpp

```diff
@@ -2932,15 +2932,24 @@ void map::collapse_at( const tripoint &p, const bool silent )
     }
 }
 
-void map::smash_items( const tripoint &p, const int power )
+void map::smash_items( const tripoint &p, const int power, const std::string &cause_message )
 {
     if( !has_items( p ) ) {
         return;
     }
 
+    // Keep track of how many items have been damaged, and what the most recent one is
+    bool item_was_damaged = false;
+    int items_damaged = 0;
+    std::string most_recently_damaged = "";
```

Comment:
You can avoid the overhead of unnecessary calls to `item::tname()` by saving a pointer to the most recently damaged item, instead of saving its name.

Reply:
I tried that, but I'm %99 percent sure that's what was causing the segmentation fault.

Reply:
That's probably because the items can be removed after being completely damaged. Keep it this way then, if there's no significant slowdown when a large amount of items are being smashed.

Reply:
On reflection, it might be faster to add a conditional statement to only set `most_recently_damaged` the first time an item is damaged.

Reply:
Changed this in the latest commit. It now assigns the name of an item to a string at most once per invocation of `map::smash_items`